pkl-core/src/main/java/org/pkl/core/ast/expression/generator/ObjectData.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.ast.expression.generator;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.collections.UnmodifiableEconomicMap;
+import org.pkl.core.ast.member.ObjectMember;
+import org.pkl.core.runtime.VmObject;
+import org.pkl.core.runtime.VmUtils;
+import org.pkl.core.util.EconomicMaps;
+
+/** Data collected by {@link GeneratorObjectLiteralNode} to generate a `VmObject`. */
+public final class ObjectData {
+  // The object's members.
+  private final EconomicMap<Object, ObjectMember> members;
+  // The frames that were active when `members` were generated.
+  // Only a subset of members have their frames stored (`GeneratorMemberNode.isFrameStored`).
+  // Frames are stored in `owner.extraStorage` and retrieved by `RestoreForBindingsNode`
+  // when members are executed.
+  private final EconomicMap<Object, MaterializedFrame> generatorFrames;
```

Comment:
This code:

```pkl
foo {
  for (i in IntSeq(1, 100)) {
    i
  }
}
```

Results in this `generatorFrames` map (in Pkl syntax), where there is really only one frame (represented as `THE_FRAME`):

```pkl
Map(
  1, THE_FRAME,
  2, THE_FRAME,
  3, THE_FRAME,
  4, THE_FRAME,
  5, THE_FRAME,
  // ...
  100, THE_FRAME
)
```

I'm not sure if this is the right model; there's many members for just this one frame. It's doing a lot of extra allocation here.

It seems quite extra to use for-generator keys as a way to look up the same materialized frame. 

Food for thought: maybe the lookup key can be a synthesized name for the for-generator. It doesn't matter too much what this name is, but it should be simple enough to use that for lookup in `RestoreForBindingsNode`.